src/frontend/src/storage/index.ts

```diff
@@ -40,69 +27,151 @@ export const getAnchors = async (): Promise<bigint[]> => {
 
 /** Set the specified anchor as used "just now" */
 export const setAnchorUsed = async (userNumber: bigint) => {
-  await withAnchors((anchors) => {
+  await withStorage((storage) => {
     const ix = userNumber.toString();
 
+    const anchors = storage.anchors;
+    const oldAnchor = anchors[ix] ?? { knownPrincipalDigests: [] };
+
     // Here we try to be as non-destructive as possible and we keep potentially unknown
     // fields
-    anchors[ix] = { ...anchors[ix], lastUsedTimestamp: nowMillis() };
-    return anchors;
+    storage.anchors[ix] = { ...oldAnchor, lastUsedTimestamp: nowMillis() };
+    return storage;
+  });
+};
+
+/** Look up an anchor by principal.
+ * In reality the anchor is looked up by principal _digest_, see `computePrincipalDigest` for
+ * more information.
+ */
+export const getAnchorByPrincipal = async ({
+  origin,
+  principal,
+}: {
+  origin: string;
+  principal: Principal;
+}): Promise<bigint | undefined> => {
+  const storage = await readStorage();
+  const anchors = storage.anchors;
+
+  const digest = await computePrincipalDigest({
+    origin,
+    principal,
+    hasher: storage.hasher,
+  });
+
+  for (const ix in anchors) {
+    const anchor: AnchorV3 = anchors[ix];
+
+    if (anchor.knownPrincipalDigests.includes(digest)) {
+      return BigInt(ix);
+    }
+  }
+
+  return;
+};
+
+/** Set the principal as "known"; i.e. from which the anchor can be "looked up" */
+export const setKnownPrincipal = async ({
+  userNumber,
+  origin,
+  principal,
+}: {
+  userNumber: bigint;
+  origin: string;
+  principal: Principal;
+}) => {
+  await withStorage(async (storage) => {
+    const ix = userNumber.toString();
+
+    const anchors = storage.anchors;
+
+    const digest = await computePrincipalDigest({
+      origin,
+      principal,
+      hasher: storage.hasher,
+    });
+
+    const oldAnchor = anchors[ix] ?? { knownPrincipalDigests: [] };
+
+    const knownPrincipalDigests = [
+      ...new Set(oldAnchor.knownPrincipalDigests.concat([digest])),
```

Comment:
I think it might be beneficial to limit the size of this array somehow.
In particular, developers that might use many deployments of II on some host (i.e. `localhost`) might get a disproportionate number of "knownPrincipals" for the anchor 10000. I.e. every redeployment of II will add a new entry for the same dapp as the salt in the canister will have changed.

Reply:
That's true! and we do that for known anchors. I'll prune this too, somehow. Will need to add a timestamp too.